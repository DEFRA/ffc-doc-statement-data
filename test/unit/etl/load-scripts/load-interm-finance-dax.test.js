const { etlConfig } = require('../../../../app/config')
const db = require('../../../../app/data')
const { loadIntermFinanceDAX } = require('../../../../app/etl/load-scripts/load-interm-finance-dax')

jest.mock('../../../../app/data', () => ({
  sequelize: {
    query: jest.fn()
  },
  etlStageLog: {
    findAll: jest.fn()
  },
  Sequelize: {
    Op: {
      gt: Symbol('gt')
    }
  }
}))

let mockTransaction
describe('loadIntermFinanceDAX', () => {
  const startDate = '2023-01-01'

  beforeEach(() => {
    mockTransaction = {}
    db.etlStageLog.findAll.mockClear()
    db.sequelize.query.mockClear()
  })

  test('should throw an error if multiple records are found', async () => {
    db.etlStageLog.findAll.mockResolvedValue([{ file: 'Finance_Dax/export.csv', idFrom: 1, idTo: 2 }, { file: 'Finance_Dax/export.csv', idFrom: 3, idTo: 4 }])

    await expect(loadIntermFinanceDAX(startDate, mockTransaction)).rejects.toThrow(
      `Multiple records found for updates to ${etlConfig.financeDAX.folder}, expected only one`
    )
  })

  test('should return if no records are found', async () => {
    db.etlStageLog.findAll.mockResolvedValue([])

    await expect(loadIntermFinanceDAX(startDate, mockTransaction)).resolves.toBeUndefined()
    expect(db.sequelize.query).not.toHaveBeenCalled()
  })

  test('should call sequelize.query with correct SQL and parameters', async () => {
    db.etlStageLog.findAll.mockResolvedValue([{ file: 'Finance_Dax/export.csv', idFrom: 1, idTo: 2 }])

    await loadIntermFinanceDAX(startDate, mockTransaction)

    expect(db.sequelize.query).toMatchSnapshot()
  })

  test('should handle errors thrown by sequelize.query', async () => {
    db.etlStageLog.findAll.mockResolvedValue([{ file: 'Finance_Dax/export.csv', idFrom: 1, idTo: 2 }])
    db.sequelize.query.mockRejectedValue(new Error('Query failed'))

    await expect(loadIntermFinanceDAX(startDate, mockTransaction)).rejects.toThrow('Query failed')
  })
})
