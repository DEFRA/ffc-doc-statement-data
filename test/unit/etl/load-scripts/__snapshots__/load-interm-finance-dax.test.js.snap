// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadIntermFinanceDAX should call sequelize.query with correct SQL and parameters 1`] = `
[MockFunction] {
  "calls": [
    [
      "
  WITH newdata AS (
    SELECT
      transdate,
      invoiceid,
      scheme::integer,
      fund,
      marketingyear::integer,
      "month",
      quarter,
      CAST(
        COALESCE(
          (SELECT CAST((value - lag) / -100.00 AS DECIMAL(10,2)) AS value 
            FROM (
              SELECT
                value,
                COALESCE(LAG(value, 1) OVER (ORDER BY S."settlementDate" ASC, S.value ASC),0) AS lag,
                S.reference
                FROM "etlStageSettlement" S 
                WHERE S."invoiceNumber" = D.invoiceid
                ORDER BY value
            ) B WHERE B.reference = D.settlementvoucher),
          lineamountmstgbp)
        AS DECIMAL(10,2)) AS "transactionAmount",
      agreementreference,
      substring(invoiceid, 2, position('Z' in invoiceid) - (position('S' in invoiceid) + 2))::integer AS "sitiInvoiceId",
      substring(invoiceid, position('Z' in invoiceid) + 1, position('V' in invoiceid) - (position('Z' in invoiceid) + 1))::integer AS "claimId",
      settlementvoucher AS "paymentRef",
      "changeType",
      recid
    FROM public."etlStageFinanceDax" D
    WHERE LENGTH(accountnum) = 10
      AND "etlId" BETWEEN :idFrom AND :idTo
      AND invoiceid LIKE 'S%Z%'
  ),
  updatedrows AS (
    UPDATE public."etlIntermFinanceDax" interm
    SET
      transdate = newdata.transdate,
      scheme = newdata.scheme,
      fund = newdata.fund,
      marketingyear = newdata.marketingyear,
      "month" = newdata."month",
      quarter = newdata.quarter,
      "transactionAmount" = newdata."transactionAmount",
      agreementreference = newdata.agreementreference,
      "sitiInvoiceId" = newdata."sitiInvoiceId",
      "claimId" = newdata."claimId",
      invoiceid = newdata.invoiceid,
      "etlInsertedDt" = NOW()
    FROM newdata
    WHERE newdata."changeType" = 'UPDATE'
      AND interm.recid = newdata.recid
    RETURNING interm.recid
  )
  INSERT INTO public."etlIntermFinanceDax" (
    transdate,
    invoiceid,
    scheme,
    fund,
    marketingyear,
    "month",
    quarter,
    "transactionAmount",
    agreementreference,
    "sitiInvoiceId",
    "claimId",
    "paymentRef",
    recid
  )
  SELECT
    transdate,
    invoiceid,
    scheme,
    fund,
    marketingyear,
    "month",
    quarter,
    "transactionAmount",
    agreementreference,
    "sitiInvoiceId",
    "claimId",
    "paymentRef",
    recid
  FROM newdata
  WHERE "changeType" = 'INSERT'
    OR ("changeType" = 'UPDATE' AND recid NOT IN (SELECT recid FROM updatedrows));
",
      {
        "raw": true,
        "replacements": {
          "idFrom": 1,
          "idTo": 2,
        },
        "transaction": {},
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
